sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala

```diff
@@ -44,39 +43,52 @@ abstract class Collect extends ImperativeAggregate {
 
   override def dataType: DataType = ArrayType(child.dataType)
 
-  override def supportsPartial: Boolean = false
-
-  override def aggBufferAttributes: Seq[AttributeReference] = Nil
-
-  override def aggBufferSchema: StructType = StructType.fromAttributes(aggBufferAttributes)
-
-  override def inputAggBufferAttributes: Seq[AttributeReference] = Nil
-
   // Both `CollectList` and `CollectSet` are non-deterministic since their results depend on the
   // actual order of input rows.
   override def deterministic: Boolean = false
 
-  protected[this] val buffer: Growable[Any] with Iterable[Any]
-
-  override def initialize(b: InternalRow): Unit = {
-    buffer.clear()
-  }
-
-  override def update(b: InternalRow, input: InternalRow): Unit = {
-    // Do not allow null values. We follow the semantics of Hive's collect_list/collect_set here.
-    // See: org.apache.hadoop.hive.ql.udf.generic.GenericUDAFMkCollectionEvaluator
-    val value = child.eval(input)
-    if (value != null) {
-      buffer += value
+  protected def _serialize(obj: Iterable[Any]): Array[Byte] = {
```

Comment:
We can generalize this to cover `Percentile`, but we may not use `T <: Growable[Any] with Iterable[Any]` but `T <: Iterable[Any]` then.